File: leetcode/rust/src/2.add-two-numbers.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn add_two_numbers(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        Self::add(l1,l2,0)
    }

    fn add(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>, carry: i32) -> Option<Box<ListNode>>{
        let mut sum = carry;
        
        let next_lhs = if let Some(nodel) = l1{
            sum += nodel.val;
            nodel.next
        }else{
            None
        };

        let next_rhs = if let Some(noder) = l2{
            sum += noder.val;
            noder.next
        } else{
            None
        };

        let mut sum_node = ListNode::new( sum % 10);

        if next_lhs != None || next_rhs != None || sum >= 10{
            sum_node.next = Self::add(next_lhs,next_rhs,sum / 10);
        }
        Some(Box::new(sum_node))

    }
}

File: leetcode/java/2.add-two-numbers.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode(int x) { val = x; } }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }

        ListNode dummy = new ListNode(-1);
        ListNode head = dummy;

        boolean carry = false;
        while (l1 != null || l2 != null || carry) {
            int n = 0;
            if (l1 != null) {
                n += l1.val;
                l1 = l1.next;
            }

            if (l2 != null) {
                n += l2.val;
                l2 = l2.next;
            }
            if (carry) {
                n++;
            }

            carry = n >= 10;
            n %= 10;

            head.next = new ListNode(n);
            head = head.next;

        }

        return dummy.next;
    }
}
